mpz-core/src/lpn.rs

```diff
@@ -0,0 +1,180 @@
+///
+/// Given a vector `x` and `e`, compute `y = Ax + e`.
+///
+/// `A` - is a binary matrix with `k` columns and `n` rows. The concrete number of `n` is determined by the input length. `A` will be generated on-the-fly.
+///
+/// `x` - is a `F_{2^128}` vector with length `k`.
+///
+/// `e` - is a `F_{2^128}` vector with length `n`.
+///
+/// Note that in the standard LPN problem, `x` is a binary vector, `e` is a sparse binary vector. The way we difined here is a more generic way in term of computing `y`.
+pub struct Lpn<const D: usize> {
+    /// The seed to generate the random sparse matrix A.
+    seed: Block,
+
+    /// The length of the secret, i.e., x.
+    k: u32,
+
+    /// A mask to optimize reduction operation.
+    mask: u32,
+}
+
+impl<const D: usize> Lpn<D> {
+    /// Create a new LPN instance.
+    pub fn new(seed: Block, k: u32) -> Self {
+        let mut mask = 1;
+        while mask < k {
+            mask <<= 1;
+            mask |= 0x1;
+        }
+        Self { seed, k, mask }
+    }
+
+    /// Compute 4 rows as a batch, this is for the `compute` function.
+    #[inline]
+    fn compute_four_rows_indep(&self, y: &mut [Block], x: &[Block], pos: usize, prp: &Prp) {
+        let mut cnt = 0u64;
+        let index: [Block; D] = std::array::from_fn(|_| {
+            let i = cnt;
+            cnt += 1;
+            Block::from(bytemuck::cast::<_, [u8; 16]>([pos as u64, i]))
+        });
+
+        let mut index = prp.permute_many_blocks(index);
+        let index = bytemuck::cast_slice_mut::<_, u32>(&mut index);
+
+        for (i, y) in y.iter_mut().enumerate().take(4) {
+            for ind in index[i * D..(i + 1) * D].iter_mut() {
+                *ind &= self.mask;
+                *ind = if *ind >= self.k { *ind - self.k } else { *ind };
+
+                *y ^= x[*ind as usize];
+            }
+        }
+    }
+
+    #[inline]
+    fn compute_one_row(&self, y: &mut [Block], x: &[Block], pos: usize, prp: &Prp) {
+        let block_size = (D + 4 - 1) / 4;
+        let mut index = (0..block_size)
+            .map(|i| Block::from(bytemuck::cast::<_, [u8; 16]>([pos as u64, i as u64])))
+            .collect::<Vec<Block>>();
+        prp.permute_block_inplace(&mut index);
+        let index = bytemuck::cast_slice_mut::<_, u32>(&mut index);
+
+        for ind in index.iter_mut().take(D) {
+            *ind &= self.mask;
+            *ind = if *ind >= self.k { *ind - self.k } else { *ind };
+            y[pos] ^= x[*ind as usize];
+        }
+    }
+
+    /// Compute `Ax + e`, writing the result in-place into `y`.
+    ///
+    /// # Arguments
+    ///
+    /// * `x` - Secret vector with length `k`.
+    /// * `y` - Error vector with length `n`, this is actually `e` in LPN.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `x.len() !=k` or `y.len() != n`.
+    pub fn compute(&self, y: &mut [Block], x: &[Block]) {
+        assert_eq!(x.len() as u32, self.k);
+        assert!(x.len() >= D);
```

Comment:
@xiangxiecrypto , shouldn't we have here be a specific algorithm which bounds the minimal amount of rows/columns depending on D?

As mentioned in https://eprint.iacr.org/2017/617
"Then, for 80-bit security and d = 10 it turns out that we will need approximately k = 182 columns and k 2 rows, while for 100-bit security we need k = 240."

Reply:
The security also depends on `k` and Hamming weight of the error vector. I will specify a new struct to contain these parameters (together with others) when I implement the COT protocol. We could simply focus on `D = 10` here.
It is only for computation purpose here, maybe I should change the name to `LpnEncoder`